presto-main/src/main/java/com/facebook/presto/execution/buffer/PagesSerde.java

```diff
@@ -0,0 +1,94 @@
+import static java.lang.Math.toIntExact;
+import static java.util.Objects.requireNonNull;
+
+public class PagesSerde
+{
+    private static final double MINIMUM_COMPRESSION_RATIO = 0.8;
+
+    private final BlockEncodingSerde blockEncodingSerde;
+    private final Optional<Compressor> compressor;
+    private final Optional<Decompressor> decompressor;
+    private final PageCompression compression;
+
+    public PagesSerde(BlockEncodingSerde blockEncodingSerde, PageCompression compression, Optional<Compressor> compressor, Optional<Decompressor> decompressor)
+    {
+        this.blockEncodingSerde = requireNonNull(blockEncodingSerde, "blockEncodingSerde is null");
+        this.compressor = requireNonNull(compressor, "compressor is null");
+        this.decompressor = requireNonNull(decompressor, "decompressor is null");
+        checkArgument(compressor.isPresent() == decompressor.isPresent(), "compressor and decompressor must both be present or both be absent");
+        this.compression = requireNonNull(compression, "compression is null");
+        checkArgument(compressor.isPresent() || compression == UNCOMPRESSED, "compressor and decompressor must be set if codec is configured");
+    }
+
+    public SerializedPage serialize(Page page)
+    {
+        SliceOutput serializationBuffer = new DynamicSliceOutput(toIntExact((page.getSizeInBytes() + 4)));
+        writeRawPage(page, serializationBuffer, blockEncodingSerde);
+
+        if (!compressor.isPresent()) {
+            return new SerializedPage(serializationBuffer.slice(), UNCOMPRESSED, page.getPositionCount(), serializationBuffer.size());
+        }
+
+        int maxCompressedLength = maxCompressedLength(serializationBuffer.size());
+        byte[] compressionBuffer = new byte[maxCompressedLength];
+        int actualCompressedLength = compressor.get().compress(serializationBuffer.slice().getBytes(), 0, serializationBuffer.size(), compressionBuffer, 0, maxCompressedLength);
+
+        if (((1.0 * actualCompressedLength) / serializationBuffer.size()) > MINIMUM_COMPRESSION_RATIO) {
+            return new SerializedPage(serializationBuffer.slice(), UNCOMPRESSED, page.getPositionCount(), serializationBuffer.size());
+        }
+
+        return new SerializedPage(
+                Slices.copyOf(Slices.wrappedBuffer(compressionBuffer, 0, actualCompressedLength)),
+                compression,
+                page.getPositionCount(),
+                serializationBuffer.size());
+    }
+
+    public Page deserialize(SerializedPage serializedPage)
+    {
+        checkArgument(serializedPage != null, "serializedPage is null");
+
+        if (!decompressor.isPresent() || serializedPage.getCompression() == UNCOMPRESSED) {
+            return readRawPage(serializedPage.getPositionCount(), serializedPage.getSlice().getInput(), blockEncodingSerde);
+        }
+
+        int uncompressedSize = serializedPage.getUncompressedSizeInBytes();
+        byte[] decompressed = new byte[uncompressedSize];
```

Comment:
I think you might want to reuse this one also.... I think the block decoders "copy" into slices so it would work.